x-pack/filebeat/input/httpjson/request_chain_helper.go

```diff
@@ -186,6 +195,36 @@ func iterateRecursive(m mapstr.M, keys []string, depth int) (interface{}, error)
 	}
 }
 
+// processExpression, splits the expression string based on the separator and looks for
+// supported keywords. If present, returns an expression array containing separated elements.
+// If no keywords are present, the expression is treated as a hardcoded value and returned
+// as a merged string which is the only array element.
+func processExpression(expression string, separator string) []string {
+	expArr := strings.Split(expression, separator)
+	startIndex := 0
+	isExpressionValue := false
+
+	for i, v := range expArr {
+		if v == "" {
+			continue
+		}
+		startIndex = i
+		if v != lastResponse && v != firstResponse && v != parentLastResponse {
+			isExpressionValue = true
+		}
+		break
+	}
+
+	var returnArr []string
+	if isExpressionValue {
+		returnArr = append(returnArr, strings.Join(expArr, separator))
+	} else {
+		returnArr = expArr[startIndex:]
+	}
+
+	return returnArr
+}
```

Comment:
I don't think that we need to generalise this for other separators and we can save work if we don't split until it's necessary. Early test is there to prevent repeated retests for cases where the expression is a literal without a leading dot (I would imagine these are more common that dot-prefixed literals).

```suggestion
func processExpression(expression string) []string {
	if !strings.HasPrefix(expression, ".") {
		return []string{expression}
	}
	switch {
	case strings.HasPrefix(expression, "."+firstResponse+"."),
		strings.HasPrefix(expression, "."+lastResponse+"."),
		strings.HasPrefix(expression, "."+parentLastResponse+"."):
		return strings.Split(expression, ".")[2:]
	default:
		return []string{expression}
	}
}
```
This has slightly different behaviour to the code above, but I think it is the less surprising of the two. Even with this, I think this func deserves a test so that the behaviour is explicitly set. The question-marked cases are things I think need consideration.
- [ ] `.first_response.foo` → ["foo"]
- [ ] `.first_response.` → ?
- [ ] `.first_response` → ?
- [ ] `.first_response.foo` → ["foo"]
- [ ] `.first_response.` → ?
- [ ] `.first_response` → ?
- [ ] `.last_response.foo` → ["foo"]
- [ ] `.last_response.` → ?
- [ ] `.last_response` → ?
- [ ] `.parent_last_response.foo` → ["foo"]
- [ ] `.literal_foo` → [".literal_foo"]
- [ ] `.literal_foo.bar` → [".literal_foo.bar"]
- [ ] `literal.foo.bar` → ["literal.foo.bar"]
- [ ] `first_response.foo` → ["first_response.foo"]

Reply:
This would work if we return **strings.Split(expression, ".")[1:]**, then it would have similar results to the previous method. 

Also in the following cases : 

.first_response. -> [first_response  ]
.first_response -> [.first_response]
.last_response. -> [last_response  ]
.last_response -> [.last_response] 

would be handled by

Reply:
Also with this change we make the '.' (dot) prefix mandatory if user wants to treat it as an expression to be evaluated and not a hardcoded value ..which is good since we have somewhat of a deterministic behaviour here.

Reply:
I think the documentation around this need more work then. It's not yet clear to me why those should be an error.

Reply:
yes will update the docs as well as introducing errors for all scenarios here .

Reply:
Can you add this test please.
```
func TestProcessExpression(t *testing.T) {
	tests := []struct {
		in   string
		want []string
	}{
		// Cursor values.
		{in: ".first_response.foo", want: []string{"first_response", "foo"}},
		{in: ".first_response.", want: []string{"first_response", ""}},
		{in: ".last_response.foo", want: []string{"last_response", "foo"}},
		{in: ".last_response.", want: []string{"last_response", ""}},
		{in: ".parent_last_response.foo", want: []string{"parent_last_response", "foo"}},
		{in: ".parent_last_response.", want: []string{"parent_last_response", ""}},

		// Literal values.
		{in: ".literal_foo", want: []string{".literal_foo"}},
		{in: ".literal_foo.bar", want: []string{".literal_foo.bar"}},
		{in: "literal.foo.bar", want: []string{"literal.foo.bar"}},
		{in: "first_response.foo", want: []string{"first_response.foo"}},
		{in: ".first_response", want: []string{".first_response"}},
		{in: ".last_response", want: []string{".last_response"}},
		{in: ".parent_last_response", want: []string{".parent_last_response"}},
	}
	for _, test := range tests {
		got := processExpression(test.in)
		assert.Equal(t, test.want, got)
	}
}
```